### ISSUE TITLE
DATA: Create `struct` for Model Name

### ISSUE THREAD

**Issue by ElbekDeveloper** [hassanhabib/OpenAI.NET#60]

**Comment by ElbekDeveloper** [comment]
![image](https://user-images.githubusercontent.com/52498074/226088274-c5bf1edc-02fc-46e5-9f94-217d63128c51.png)
![image](https://user-images.githubusercontent.com/52498074/226088283-f1ce5cac-58cf-4a72-b16b-6509658d7562.png)
These are GPT-3 Models
Here is how we want:
![image](https://user-images.githubusercontent.com/52498074/226088301-a50a82fa-d1e5-41af-b870-04ed33f6d652.png)
That's the example implementation:
![image](https://user-images.githubusercontent.com/52498074/226088329-a31c688e-bfbc-48e4-a3f6-a31e79a4e9a4.png)

**Comment by jodendaal** [comment]
I think I nice way of doing this would be to use the implicit operator something like this, this allows us to define our own ModelTypes which works with intellisense, but also allows any custom model string the consumer may want to pass in.

`internal class Program
    {
        static void Main(string[] args)
        {
            var textCompletion = new TextCompletionRequest(Model.Whisper1);
            textCompletion.Model = Model.TextDavinci003;

            var textCompletionCustom = new TextCompletionRequest("my-custom-model");
            textCompletion.Model = "my-custom-model";
        }
    }

    public class TextCompletionRequest
    {
        public TextCompletionRequest(Model model)
        { 
        
        }

        public Model Model { get; set; }
    }

    public class Model
    {
        /// <summary>
        /// text-davinci-003
        /// </summary>
        public static readonly Model TextDavinci003  = new Model("text-davinci-003");

        /// <summary>
        /// GPT35Turbo 
        /// </summary>
        public static readonly Model GPT35Turbo = new Model("gpt-3.5-turbo");

        /// <summary>
        /// whisper-1
        /// </summary>
        public static readonly Model Whisper1 = new Model("whisper-1");


        private readonly string _value;

        private Model(string value)
        {
            _value = value;
        }

        public override string ToString()
        {
            return _value;
        }

        public static implicit operator string(Model action)
        {
            return action._value;
        }

        public static implicit operator Model(string modelName)
        {
            return new Model(modelName);
        }
    }`

**Comment by StefH** [comment]
Building on the answer from @jodendaal, you can use a source generator like [StronglyTypedId](https://github.com/meziantou/Meziantou.Framework/tree/main/src/Meziantou.Framework.StronglyTypedId) to generate the code for the struct.

The only thing you'll need to add is those implicit operators, example looks like:
``` c#
[StronglyTypedId(typeof(string))]
public partial struct Model
{
    /// <summary>
    /// text-davinci-003
    /// </summary>
    public static readonly Model TextDavinci003 = Parse("text-davinci-003");

    /// <summary>
    /// GPT35Turbo 
    /// </summary>
    public static readonly Model GPT35Turbo = Parse("gpt-3.5-turbo");

    /// <summary>
    /// whisper-1
    /// </summary>
    public static readonly Model Whisper1 = Parse("whisper-1");

    public static implicit operator string(Model action)
    {
        return Guard.NotNullOrEmpty(action.Value);
    }

    public static implicit operator Model(string modelName)
    {
        return Model.Parse(Guard.NotNullOrEmpty(modelName));
    }
}
```

Using it is the same:
``` c#
var textCompletion = new TextCompletionRequest(Model.Whisper1);
textCompletion.Model = Model.TextDavinci003;

var textCompletionCustom = new TextCompletionRequest("my-custom-model");
textCompletion.Model = "my-custom-model";
```

**Comment by jodendaal** [comment]
Just updated to include a ModelConvertor, This will work with Newtonsoft.Json which I think you are using.

**Comment by StefH** [comment]
Note that that StronglyTypedId package also automatically generates all typeconverters.

**Comment by StefH** [comment]
@hassanhabib 
And a totally different approach would be to use [AnyOf](https://github.com/StefH/AnyOf) which allows for two (or more) defined types.
So in this case a string and a Model.

``` c#
var any = AnyOf<Model, string>();

any = "str";
any = new Model();
```

(But maybe this will not help in this situation, however, maybe useful for other scenarios)

**Comment by hassanhabib** [comment]
> @hassanhabib 
> And a totally different approach would be to use [AnyOf](https://github.com/StefH/AnyOf) which allows for two (or more) defined types.
> So in this case a string and a Model.
> 
> ``` c#
> var any = AnyOf<Model, string>();
> 
> any = "str";
> any = new Model();
> ```
> 
> (But maybe this will not help in this situation, however, maybe useful for other scenarios)

Wow I didn't know that existed.
Won't help here. We need auto complete like enums.
But good to know. Thank you for always sharing ❤️

**Comment by jodendaal** [comment]
Having string values on Enums for reference, not appropriate for this solution but something I have used before, all values where known though. Extended the solution linked to be able to parse from string value back to Enum. I'm not a huge fan of it but it works in some scenarios.

https://stackoverflow.com/a/630900/2207784

**Comment by mabroukmahdhi** [comment]
> @hassanhabib And a totally different approach would be to use [AnyOf](https://github.com/StefH/AnyOf) which allows for two (or more) defined types. So in this case a string and a Model.
> 
> ```cs
> var any = AnyOf<Model, string>();
> 
> any = "str";
> any = new Model();
> ```
> 
> (But maybe this will not help in this situation, however, maybe useful for other scenarios)

In this case our model will be depending on a 3d-party library.. not nice 🗡️ 

And as @hassanhabib  said, we need auto complete

**Comment by StefH** [comment]
@hassanhabib / @ElbekDeveloper 

Another approach to make using and initializing the models is by using a [FluentBuilder](https://github.com/StefH/FluentBuilder).

So instead of this:
``` c#
var inputCompletion = new Completion
{
	Request = new CompletionRequest
	{
		Prompts = new string[] {"Human: Hello!"}
		Model = "text-davinci-003"
	};
}
```

You can use this:
``` c#
var inputCompletion = new Completion
{
	Request = new CompletionRequest
	{
		Prompts = new string[] {"Human: Hello!"}
		Model = "text-davinci-003"
	};
}
```

You can offer the users an interface like this:
``` c#
// using builder with a string
var inputCompletionUsingBuilder1 = new CompletionBuilder()
    .WithRequest(requestBuilder => requestBuilder
        .WithPrompt("hello")
        .WithModel("my model")
        .Build()
    )
    .Build();


// using builder with a struct
var inputCompletionUsingBuilder2 = new CompletionBuilder()
    .WithRequest(requestBuilder => requestBuilder
        .WithPrompt("hello")
        .WithModel(Model.TextDavinci003)
        .Build()
    )
    .Build();

// using builder with a enum
var inputCompletionUsingBuilder3 = new CompletionBuilder()
    .WithRequest(requestBuilder => requestBuilder
        .WithPrompt("hello")
        .WithModel(ModelAsEnum.TextDavinci003)
        .Build()
    )
    .Build();
```

See my example PR here:
https://github.com/hassanhabib/OpenAI.NET/pull/73

(Note that in this example I've created the builders in my test console app, however you can also annotate your model classes with the AutoGenerateBuilder attribute and create those partial builder classes in your main project.

**Comment by diogod3** [comment]
https://github.com/dotnet/csharplang/issues/2849

This would solve our problem.

**Comment by jodendaal** [comment]
> [dotnet/csharplang#2849](https://github.com/dotnet/csharplang/issues/2849)
> 
> This would solve our problem.

Could you point me to the bit you are referring too, seems like a language feature that doesn't exist yet?

**Comment by diogod3** [comment]
Exactly, I was saying that "string enums" would solve our problem, but unfortunately that feature doesn't exist.
